### PR TITLE
RGB image

### PR DESCRIPTION
**Pull request by davidpiuva**
Adds RGB and RGBA images for CPU and OpenGL.

### CODE REVIEW

**Comment by marcusnaslund** [comment]
#1393 will be evaluated right after lunch.

**Comment by davidpiuva** [comment]
Rebased.

**Comment by davidpiuva** [comment]
@emilwestergren Peer review?

**Comment by marcusnaslund** [comment]
I still expect tests for the new `Raster` classes. Should be simple enough to adapt the same tests as those for the `Bgr` versions.

**Comment by davidpiuva** [comment]
If BGR will be removed then I can rewrite them to use RGB. It will be a huge pull request.

**Comment by emilwestergren** [comment]
Looks good, but I agree with @marcusnaslund, can you copy the tests for BGR and then we can simply remove all BGR files in a single PR after this?

**Comment by marcusnaslund** [comment]
In comparison to everything else I see nothing huge about the `RasterBgr` tests.

**Comment by davidpiuva** [comment]
Actually removing BGR should probably go in a separate pull request so that it can be reverted in the future or at least used as a reference if we suddenly want it back.

**Comment by emilwestergren** [comment]
Yes

**Comment by davidpiuva** [comment]
My "Rgb image adapt" pull requests are removing around half of what is using BGR right now.

**Comment by emilwestergren** [comment]
So:
1. Include tests for RGB in this PR
2. Create PR that removes any use of BGR
3. Create PR that removes BGR

**Comment by davidpiuva** [comment]
So I just copy the files and replace BGR with RGB for its own tests while leaving all unrelated tests using BGR?

**Comment by emilwestergren** [comment]
Yes

**Comment by davidpiuva** [comment]
One of the tests can not be implemented for RGB now because saving of PNG images is hard coded for BGR. I have to make some workaround by flipping channels on a copy in order to delete BGR.

**Comment by davidpiuva** [comment]
The thread pool test is still randomly crashing.

**Comment by davidpiuva** [comment]
Now I just need to break the infinite recursion by calling the base save function.
```
save: override func (filename: String) -> Int {
	bgra := this redBlueSwapped()
	result := bgra save(filename)
	bgra free()
	result
}
```

**Comment by sebastianbaginski** [comment]
https://github.com/magic-lang/ooc-kean/issues/1409 it happened to my PRs too few days ago

**Comment by davidpiuva** [comment]
I failed to override save when using super.
The generated virtual function table shows that RasterPacked's save is still called causing the tests to fail. `.save = (void*) RasterPacked__RasterPacked_save_impl`

**Comment by davidpiuva** [comment]
When commenting out the call to super, everything works.
`.save = (void*) RasterRgb__RasterRgb_save_impl`

**Comment by davidpiuva** [comment]
I might have to duplicate the save function from Packed until the compiler is fixed.

**Comment by davidpiuva** [comment]
Updated.

File: source/draw/RasterRgb.ooc [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	}
+	_createCanvas: override func -> Canvas { RasterRgbCanvas new(this) }
+
+	operator [] (x, y: Int) -> ColorRgb { this isValidIn(x, y) ? ((this buffer pointer + y * this stride) as ColorRgb* + x)@ : ColorRgb new(0, 0, 0) }
+	operator []= (x, y: Int, value: ColorRgb) { ((this buffer pointer + y * this stride) as ColorRgb* + x)@ = value }
+
+	open: static func (filename: String) -> This {
+		This convertFrom(RasterBgr open(filename))
+	}
+	savePacked: func (filename: String) -> Int {
```
Comment (davidpiuva) [inline source/draw/RasterRgb.ooc]: This will be removed once super works.

File: source/draw/RasterRgba.ooc [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+			(top * (left * topLeft green + (1 - left) * topRight green) + (1 - top) * (left * bottomLeft green + (1 - left) * bottomRight green)),
+			(top * (left * topLeft blue + (1 - left) * topRight blue) + (1 - top) * (left * bottomLeft blue + (1 - left) * bottomRight blue)),
+			(top * (left * topLeft alpha + (1 - left) * topRight alpha) + (1 - top) * (left * bottomLeft alpha + (1 - left) * bottomRight alpha))
+		)
+	}
+
+	open: static func (filename: String) -> This {
+		This convertFrom(RasterBgra open(filename))
+	}
+	savePacked: func (filename: String) -> Int {
```
Comment (davidpiuva) [inline source/draw/RasterRgba.ooc]: This will be removed once super works.

**Comment by davidpiuva** [comment]
The generated code seems to be completely random for the same input. Probably lazy compilation causing it again.

**Comment by davidpiuva** [comment]
@marcusnaslund All tests passed.

**Comment by marcusnaslund** [comment]
Conflicts.

**Comment by davidpiuva** [comment]
Made interactive rebase with new commits to handle version conflict.

**Comment by davidpiuva** [comment]
@fredrikbryntesson `SynchronizedResourceTest` crashed.

**Comment by davidpiuva** [comment]
@fredrikbryntesson All tests passed.

**Comment by davidpiuva** [comment]
@sebastianbaginski Peer review the last added commit?

File: source/draw/RasterRgb.ooc [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	operator []= (x, y: Int, value: ColorRgb) { ((this buffer pointer + y * this stride) as ColorRgb* + x)@ = value }
+
+	open: static func (filename: String) -> This {
+		x, y, imageComponents: Int
+		requiredComponents := 3
+		data := StbImage load(filename, x&, y&, imageComponents&, requiredComponents)
+		buffer := ByteBuffer new(data as Byte*, x * y * requiredComponents, true)
+		result := This new(buffer, IntVector2D new(x, y))
+		result swapRedBlue()
+		result
```
Comment (sebastianbaginski) [inline source/draw/RasterRgb.ooc]: Maybe `This convertFrom(RasterRgba open(filename))`

Comment (davidpiuva) [inline source/draw/RasterRgb.ooc]: A function returning a raw buffer from a file might be able to give the correct `requiredComponents` from the beginning.

Comment (sebastianbaginski) [inline source/draw/RasterRgb.ooc]: It does not matter, this is only for debugging / testing. You were complaining about code bloat in draw module so just use a function that already exists.

Comment (davidpiuva) [inline source/draw/RasterRgb.ooc]: So it is okay if it might crash?

Comment (sebastianbaginski) [inline source/draw/RasterRgb.ooc]: You can of course add this function if you want, but this PR is quite big already.

Comment (sebastianbaginski) [inline source/draw/RasterRgb.ooc]: What might crash ?

Comment (davidpiuva) [inline source/draw/RasterRgb.ooc]: Saying that the image has 4 dimensions when it has only 3. I don't know how the exception is handled in the external library.

Comment (sebastianbaginski) [inline source/draw/RasterRgb.ooc]: If you ask for 4 and image has only 3 then you will get only 3, `imageComponents` returns the value from file.

**Comment by sebastianbaginski** [comment]
OK just leave this as it is for now, I guess we should clean this up anyway.

**Comment by sebastianbaginski** [comment]
Great, PR done.

**Comment by davidpiuva** [comment]
@fredrikbryntesson All checks passed again.

File: source/draw/RasterRgb.ooc [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			rgb = RasterRgb convertFrom(image as RasterImage)
+		else
+			Debug error("Unsupported image type in RgbRasterCanvas draw")
+		this _resizePacked(rgb buffer pointer as ColorRgb*, rgb, source, destination)
+		if (rgb != image)
+			rgb referenceCount decrease()
+	}
+}
+
+RasterRgb: class extends RasterPacked {
```
Comment (davidpiuva) [inline source/draw/RasterRgb.ooc]: A copy of BGR with loading and saving changed.

File: source/draw/RasterRgba.ooc [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+	target ::= this _target as RasterRgba
+	init: func (image: RasterRgba) { super(image) }
+	_drawPoint: override func (x, y: Int) {
+		position := this _map(IntPoint2D new(x, y))
+		if (this target isValidIn(position x, position y))
+			this target[position x, position y] = this target[position x, position y] blend(this pen alphaAsFloat, this pen color)
+	}
+}
+
+RasterRgba: class extends RasterPacked {
```
Comment (davidpiuva) [inline source/draw/RasterRgba.ooc]: A copy of BGRA with loading and saving changed.

File: source/draw/RasterRgba.ooc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	}
+	savePacked: func (filename: String) -> Int {
+		file := File new(filename)
+		folder := file parent . mkdirs() . free()
+		file free()
+		StbImage writePng(filename, this size x, this size y, this bytesPerPixel, this buffer pointer, this size x * this bytesPerPixel)
+	}
+	save: override func (filename: String) -> Int {
+		bgra := this redBlueSwapped()
+		result := bgra savePacked(filename)
```
Comment (davidpiuva) [inline source/draw/RasterRgba.ooc]: Calling super failed to generate correct C code so the code was duplicated.

File: source/draw/RasterRgb.ooc [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	savePacked: func (filename: String) -> Int {
+		file := File new(filename)
+		folder := file parent . mkdirs() . free()
+		file free()
+		StbImage writePng(filename, this size x, this size y, this bytesPerPixel, this buffer pointer, this size x * this bytesPerPixel)
+	}
+	save: override func (filename: String) -> Int {
+		bgr := this redBlueSwapped()
+		result := bgr savePacked(filename)
```
Comment (davidpiuva) [inline source/draw/RasterRgb.ooc]: Calling super failed to generate correct C code so the code was duplicated.

**Pull request merged**